geth/txqueue/txqueue.go

```diff
@@ -38,8 +38,7 @@ var (
 
 // TxQueue is capped container that holds pending transactions
 type TxQueue struct {
-	transactions  map[common.QueuedTxID]*common.QueuedTx
-	mu            sync.RWMutex // to guard transactions map
+	transactions  *transactions
```

Comment:
Maybe try this https://golang.org/pkg/sync/#Map instead?

Reply:
I think this solution is enought for issue. We could think about sync.map if any performance issue occures in future.

Reply:
I am not saying this because of performance but because of additional 59 lines of code that we need to support. Also, what was wrong with the previous solution? It's the same but you just moved it to a separate struct.

Reply:
This code has a history: it was writen, rewrited, many tests were added, but it still has races. For my look it means that current approach doesn't work wery well.
If we want to deligate thread safaty to client we'll gain even more support in future to prevent bugs.

Reply:
The problem I see is that now `TxQueue` is not thread-safe explicitly (methods not guarded with a mutex, neither general purpose mutex in `TxQueue`). Implicitly it is because `transactions` field is thread-safe and the rest is channels. However, when someone decides to extend `TxQueue` with a new field or method, that person needs to be aware of this fact and the same work needs to be done: wrap a field in a struct with a mutex. On top of that, what will happen if changing two fields needs to be atomic? You can't do that with two mutexes because it won't be atomic, so you need to introduce a general purpose mutex anyway.

I can somehow justify this approach regarding `QueuedTx` because that struct is used in many places, but here it does not make sense.

Reply:
If someone will want to change two fields at the same time and atimic he'll write new code. At the present moment we won't need that and dont need to extend the task.
This issue is only about current data races. The creteria is passed tests with 'race' flag.

Could you provide data race scenario for TxQueue? I'll check it and fix if it exists.

Reply:
Not sure about performance here, `sync.Map` could be interesting. Only interested in avoiding premature optimisation and aspiring less code.

Reply:
The only problem with `sync.Map` is that it stores all values as `interface{}` so type conversion is required. Thus, I would stay with:
```go
transactions  map[common.QueuedTxID]*common.QueuedTx
mu            sync.RWMutex // to guard transactions map
```

Reply:
@themue I believe on mobile devices sync.map could be much slower than map+mutex.

Reply:
@JekaMas Maybe, don't exactly know. Only asking how often this map is accessed in which frequency and how much percent of a typical usage profile the code accesses the map? Often enough seen optimisation efforts for code that's only called in a very low frequency. Must not be that way here, only to explain my thoughts.

Reply:
Performance of accessing a map value (using `sync.Map` or map and mutex) is negligible compared to e.g. processing a single Whisper message. We are not in the position where this kind of stuff should be considered. Readable and easy to understand code is much more important. And I think that this change made the code less readable and did not bring any value.

Also, after reading `sync.Map` docs, it is not really built for our use case anyway.